File: internal/calendar/calendar_test.go
package calendar

import (
	"testing"
	"time"
)

func seoul(t *testing.T) Zone {
	t.Helper()
	z, err := Named("Asia/Seoul")
	if err != nil {
		t.Fatal(err)
	}
	return z
}

func la(t *testing.T) Zone {
	t.Helper()
	z, err := Named("America/Los_Angeles")
	if err != nil {
		t.Fatal(err)
	}
	return z
}

func TestDayInstantKST(t *testing.T) {
	// 2026-08-18 01:00 KST is stored as 2026-08-17T16:00:00.000Z.
	day, ok := Day("2026-08-17T16:00:00.000Z", Instant, seoul(t))
	if !ok || day != "2026-08-18" {
		t.Fatalf("got %q ok=%v", day, ok)
	}
	utc, ok := Day("2026-08-17T16:00:00.000Z", Instant, UTC())
	if !ok || utc != "2026-08-17" {
		t.Fatalf("utc day %q ok=%v", utc, ok)
	}
}

func TestInRangeKSTCreatedFrom(t *testing.T) {
	if !InRange("2026-08-17T16:00:00.000Z", Instant, "2026-08-18", "", seoul(t)) {
		t.Fatal("KST 01:00 must match created_from=2026-08-18")
	}
	if InRange("2026-08-17T16:00:00.000Z", Instant, "2026-08-18", "", UTC()) {
		t.Fatal("UTC calendar day is the 17th; must not match from=18")
	}
}

func TestDateOnlyNotUTCShifted(t *testing.T) {
	day, ok := Day("2026-08-20", Date, la(t))
	if !ok || day != "2026-08-20" {
		t.Fatalf("date-only must stay 2026-08-20, got %q ok=%v", day, ok)
	}
	// Same string passed as Instant must not go through UTC midnight either.
	day, ok = Day("2026-08-20", Instant, la(t))
	if !ok || day != "2026-08-20" {
		t.Fatalf("date-only instant must stay 2026-08-20, got %q ok=%v", day, ok)
	}
}

func TestInRangeEmptyRaw(t *testing.T) {
	if InRange("", Instant, "2026-08-18", "", seoul(t)) {
		t.Fatal("empty timestamp with a lower bound must fail")
	}
	if !InRange("", Instant, "", "", seoul(t)) {
		t.Fatal("empty timestamp with no lower bound must pass")
	}
}

func TestExplainReportsDecision(t *testing.T) {
	d := Explain("2026-08-17T16:00:00.000Z", Instant, seoul(t))
	if !d.OK || d.CalendarDay != "2026-08-18" || d.Kind != "instant" || d.Zone != "Asia/Seoul" {
		t.Fatalf("%+v", d)
	}
	d = Explain("2026-08-20", Date, la(t))
	if !d.OK || d.CalendarDay != "2026-08-20" || d.Kind != "date" {
		t.Fatalf("%+v", d)
	}
}

func TestStartOfWeekMonday(t *testing.T) {
	// Wednesday 2026-08-19 01:00 KST → Monday 2026-08-17.
	now := time.Date(2026, 8, 18, 16, 0, 0, 0, time.UTC)
	got := StartOfWeekMonday(now, seoul(t))
	if got != "2026-08-17" {
		t.Fatalf("got %s", got)
	}
}

func TestFormatDayUsesZone(t *testing.T) {
	ts := time.Date(2026, 8, 17, 16, 0, 0, 0, time.UTC)
	if got := FormatDay(ts, seoul(t)); got != "2026-08-18" {
		t.Fatalf("got %s", got)
	}
}

File: web/src/lib/calendar.test.ts
import { describe, expect, test } from 'vitest'
import {
  calendarDay,
  explain,
  formatAbs,
  inRange,
  startOfWeekMonday,
  utcZone,
  zoneNamed,
} from './calendar'

const seoul = zoneNamed('Asia/Seoul')
const la = zoneNamed('America/Los_Angeles')

describe('calendar owner (aligned with internal/calendar)', () => {
  test('instant 01:00 KST is calendar day 18 in Seoul, 17 in UTC', () => {
    expect(calendarDay('2026-08-17T16:00:00.000Z', 'instant', seoul)).toBe('2026-08-18')
    expect(calendarDay('2026-08-17T16:00:00.000Z', 'instant', utcZone())).toBe('2026-08-17')
  })

  test('created_from=2026-08-18 includes the KST 01:00 instant only in Seoul', () => {
    expect(inRange('2026-08-17T16:00:00.000Z', 'instant', '2026-08-18', null, seoul)).toBe(true)
    expect(inRange('2026-08-17T16:00:00.000Z', 'instant', '2026-08-18', null, utcZone())).toBe(
      false,
    )
  })

  test('date-only duedate is not shifted in America/Los_Angeles', () => {
    expect(calendarDay('2026-08-20', 'date', la)).toBe('2026-08-20')
    expect(calendarDay('2026-08-20', 'instant', la)).toBe('2026-08-20')
    // Date.parse("2026-08-20") is UTC midnight → 19th in LA. Owner must not.
    const parsed = new Date('2026-08-20')
    const laDay = new Intl.DateTimeFormat('en-CA', {
      timeZone: 'America/Los_Angeles',
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
    }).format(parsed)
    expect(laDay).toBe('2026-08-19')
  })

  test('empty raw: pass only when there is no lower bound', () => {
    expect(inRange(null, 'instant', '2026-08-18', null, seoul)).toBe(false)
    expect(inRange(null, 'instant', null, null, seoul)).toBe(true)
  })

  test('explain reports the decision', () => {
    expect(explain('2026-08-17T16:00:00.000Z', 'instant', seoul)).toEqual({
      raw: '2026-08-17T16:00:00.000Z',
      kind: 'instant',
      zone: 'Asia/Seoul',
      calendar_day: '2026-08-18',
      ok: true,
    })
    expect(explain('2026-08-20', 'date', la).calendar_day).toBe('2026-08-20')
  })

  test('startOfWeekMonday in KST: Wed 08-19 01:00 → Mon 08-17', () => {
    const now = new Date('2026-08-18T16:00:00.000Z')
    expect(startOfWeekMonday(now, seoul)).toBe('2026-08-17')
  })

  test('formatAbs of a date-only string does not emit a time', () => {
    const shown = formatAbs('2026-08-20', 'date', la, 'en-US')
    expect(shown).toMatch(/08/)
    expect(shown).toMatch(/20/)
    expect(shown).not.toMatch(/:/)
  })
})
